annotated.go

```diff
@@ -89,3 +90,268 @@ func (a Annotated) String() string {
+	annotatedResult := []reflect.StructField{{
+		Name:      "Out",
+		Type:      reflect.TypeOf(Out{}),
+		Anonymous: true,
+	}}
+
+	errInterface := reflect.TypeOf((*error)(nil)).Elem()
+	numAnnotated := 0
+
+	for i := 0; i < fType.NumOut(); i++ {
+		// guard against error results
+		if fType.Out(i).Implements(errInterface) {
+			continue
+		}
+		structField := reflect.StructField{
+			Name: fmt.Sprintf("Field%d", numAnnotated),
+			Type: fType.Out(i),
+		}
+		if numAnnotated < len(rt.tags) {
+			structField.Tag = reflect.StructTag(rt.tags[numAnnotated])
+		}
+		numAnnotated++
+		annotatedResult = append(annotatedResult, structField)
+	}
+
+	resultTypes := []reflect.Type{reflect.StructOf(annotatedResult)}
+	return resultTypes
+}
+
+// ResultTags is an Annotation that annotates the result(s) of a function.
+// When multiple tags are specified, each tag is mapped to the corresponding
+// positional result.
+func ResultTags(tags ...string) Annotation {
+	r := resultTags{tags}
+	return r
+}
+
+// Annotate lets you annotate a function's paramter and returns with tags
+// without you having to declare separate struct definitions for them.
+//
+// For example,
+//   func NewGateway(ro, rw *db.Conn) *Gateway { ... }
+//   fx.Provide(
+//     fx.Annotate(
+//       NewGateway,
+//       fx.ParamTags(`name:"ro" optional:"true"`, `name:"rw"`),
+//       fx.ResultTags(`name:"foo"`),
+//     ),
+//   )
+//
+// Is equivalent to,
+//
+//  type params struct {
+//    fx.In
+//
+//    RO *db.Conn `name:"ro" optional:"true"`
+//    RW *db.Conn `name:"rw"`
+//  }
+//
+//  type result struct {
+//    fx.Out
+//
+//    GW *Gateway `name:"foo"`
+//   }
+//
+//   fx.Provide(func(p params) result {
+//     return result{GW: NewGateway(p.RO, p.RW)}
+//   })
+//
+// If a single annotation appears multiple times, all annotations
+// will be ignored.
+//
+// For example,
+//
+//  fx.Provide(
+//    fx.Annotate(
+//      NewGateWay,
+//      fx.ParamTags(`name:"ro" optional:"true"`),
+//      fx.ParamTags(`name:"rw"),
+//      fx.ResultTags(`name:"foo"`)
+//    )
+//  )
+//
+// is considered an invalid usage and will not apply any of the
+// Annotations to NewGateway.
+//
+// If more tags are given than the number of parameters/results, only
+// the ones up to the number of parameters/results will be applied.
+func Annotate(f interface{}, anns ...Annotation) interface{} {
+	fVal := reflect.ValueOf(f)
+	fType := fVal.Type()
+	numIn := fType.NumIn()
+	numOut := fType.NumOut()
+
+	var annotations annotations
+	var ins []reflect.Type
+	var outs []reflect.Type
+
+	annotatedIn := false
+	annotatedOut := false
+
+	for _, ann := range anns {
+		if e := ann.apply(&annotations); e != nil {
+			return annotationError{
+				target: f,
+				err:    e,
+			}
+		}
+		if pTags, ok := ann.(paramTags); ok {
+			ins = pTags.getAnnotatedType(fType)
+			annotatedIn = true
+		}
+		if rTags, ok := ann.(resultTags); ok {
+			outs = rTags.getAnnotatedType(fType)
+			annotatedOut = true
+		}
```

Comment:
I may not fully grok this yet, but I think we have some redundancy in the Annotation interface here.
We have two methods:

- apply: fills an annotations struct with information similar to [how functional options usually work](https://github.com/uber-go/guide/blob/master/style.md#functional-options)
- getAnnotatedType: contains behavior based on that single option

I feel like it should be one or the other? We should either have just getAnnotatedType, and that holds the behavior, or we have apply fill the annotations struct above, and then we build the annotated type out of the information recorded in the annotations struct. As it stands, the annotations struct doesn't serve much purpose otherwise.